### PR TITLE
[Unity] Multi-device support for Relax

### PR DESCRIPTION
**Pull request by yongwww**
This PR is specifically for the P1 & P2 of #15101.

### CODE REVIEW

**Comment by yongwww** [comment]
cc: @tqchen @psrivas2 @junrushao @csullivan @slyubomirsky

File: src/relay/backend/contrib/cmsisnn/target.cc [hunk 1]
```diff
@@ -31,13 +31,14 @@ namespace contrib {
...
 tvm::transform::Pass RelayToTIR();
+using FTVMTIRToRuntime = runtime::TypedPackedFunc<runtime::Module(IRModule, Target)>;
 runtime::Module TIRToRuntime(IRModule mod, Target target);
 
 TVM_REGISTER_TARGET_KIND("cmsis-nn", kDLCPU)
     .add_attr_option<Array<String>>("mattr")
     .add_attr_option<String>("mcpu")
     .add_attr_option<Bool>("debug_last_error")
-    .set_attr<FTVMRelayToTIR>(tvm::attr::kRelayToTIR, RelayToTIR())
+    .set_attr<tvm::transform::Pass>(tvm::attr::kRelayToTIR, RelayToTIR())
```
Comment (tqchen) [inline src/relay/backend/contrib/cmsisnn/target.cc:41]: these changes maybe should goto main?

Comment (tqchen) [reply src/relay/backend/contrib/cmsisnn/target.cc:41]: Maybe we don't need these changes

File: include/tvm/relax/struct_info.h [hunk 2]
```diff
@@ -180,17 +184,20 @@ class TensorStructInfoNode : public StructInfoNode {
...
     v->Visit("dtype", &dtype);
+    v->Visit("vdevice", &vdevice);
     v->Visit("ndim", &ndim);
     v->Visit("span", &span);
   }
 
   bool SEqualReduce(const TensorStructInfoNode* other, SEqualReducer equal) const {
-    return equal(shape, other->shape) && equal(ndim, other->ndim) && equal(dtype, other->dtype);
+    return equal(shape, other->shape) && equal(ndim, other->ndim) &&
+           equal(vdevice, other->vdevice) && equal(dtype, other->dtype);
```
Comment (tqchen) [inline include/tvm/relax/struct_info.h:194]: This changes the json format, would be good to add a unity updater here https://github.com/apache/tvm/blob/main/python/tvm/ir/json_compact.py#L60 to upgrade the json so when we load older versions they still work

File: src/relay/backend/contrib/cmsisnn/target.cc (hunk 1 above)
Comment (yongwww) [reply src/relay/backend/contrib/cmsisnn/target.cc]: I am planning to submit my update for the Target to the main branch. In the current Target implementation, there is a dependency on IRModule : `target.h -> target_kind.h - > transform.h - > module.h`. In this pr, we introduced Target as a member of VDevice, and VDevice could be part of the global_infos of IRModule, leading to the dependency `module.h->target.h`", resulting in circular dependencies. My changes are trying to resolve this issue by removing the dependency of IRModule on Target.

File: include/tvm/relax/struct_info.h (hunk 2 above)
Comment (yongwww) [reply include/tvm/relax/struct_info.h:194]: [the updater](https://github.com/apache/tvm/pull/15447/files#diff-1a7f4665eeaa8ac3b32bb983b901e52eee4a2e38661c507e1902bd6c46ce7785R60-R72) was added in the new commit. Currently we don't have version management for Unity branch, I used `0.13` as the from_version, `0.14` as the to_version after the upgrade in the updater.

**Review by tqchen — APPROVED** [review]

**Comment by tqchen** [comment]
Thank you @yongwww ! this is merged

**Pull request merged**